### PR TITLE
BUG: Don't warn if default conflicts with dialect

### PR DESCRIPTION
**Pull request by gfyoung**
xref #23761

Also introduces a context manager for setting CSV dialect.

### CODE REVIEW

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -769,18 +787,27 @@ def __init__(self, f, engine=None, **kwds):
...
+                # Don't warn if the default parameter was passed in,
+                # even if it conflicts with the dialect (gh-23761).
+                if provided != parser_default and provided != dialect_val:
+                    msg = ("Conflicting values for '{param}': '{val}' was "
+                           "provided, but the dialect specifies '{diaval}'. "
+                           "Using the dialect-specified value.".format(
+                               param=param, val=provided, diaval=dialect_val))
+
+                    # Annoying corner case for not warning about
+                    # conflicts between dialect and delimiter parameter.
```
Comment (jreback) [inline pandas/io/parsers.py]: is there anyway to do this check here, but avoid having all of 'sep_override' stuff. I know this is a PITA.

Comment (gfyoung) [reply pandas/io/parsers.py]: PITA for sure.  Unfortunately, I couldn't come up with anything simpler.  The aliasing between `sep` and `delimiter` makes this difficult to simplify compared to other parameters.  Some sort of special casing has to come through.

I did try modifying `_parser_defaults`, but multiple `read_*` functions rely on the defaults being as they are (i.e. they are test failures).  Thus, I found it more prudent to localize the change to just `read_csv` and not touch other read methods for just one parameter.

As the IO interface cleans up more, we probably will find a better way to do this, but as it stands, I'm willing to stick by this for now.

**Comment by jreback** [comment]
thanks!

File: pandas/tests/io/parser/test_dialect.py [hunk 2]
```diff
@@ -75,17 +76,60 @@ class InvalidDialect(object):
...
+
+    expected = DataFrame({"a": [1], "b": [2]})
+    data = "a:b\n1:2"
+
+    warning_klass = None
+    kwds = dict()
+
+    # arg=None tests when we pass in the dialect without any other arguments.
+    if arg is not None:
+        if "value" == "dialect":  # No conflict --> no warning.
```
Comment (MarcoGorelli) [inline pandas/tests/io/parser/test_dialect.py:96]: presumably this was mean to be `if value == "dialect"`? fortunately pylint has caught this 😄 https://github.com/pandas-dev/pandas/pull/49502/files

**Pull request merged**